File: apps/android/app/src/main/java/dev/edifierctrl/app/session/SessionModels.kt
package dev.edifierctrl.app.session

import java.util.Locale

/** 在 Compose 页面与应用会话 worker 之间共享的不可变数据. */
data class HeadphoneProfile(
    val id: String,
    val displayName: String,
    val capabilities: Set<String>,
    val maxNameLen: Int,
    val serviceUuid: String?,
) {
    fun supports(feature: String): Boolean = capabilities.contains(feature)
}

data class HeadphoneDevice(
    val address: String,
    val name: String,
    val kind: String = "rfcomm",
    val serviceUuid: String? = null,
) {
    val displayName: String
        get() = name.ifBlank { address }
}

data class GroupPeer(
    val id: String,
    val host: String,
    val holding: String?,
    val platform: String,
    val appVersion: String,
    val canAudio: Boolean,
) {
    val displayName: String
        get() = host.ifBlank { id }
}

data class ActivityEntry(
    val id: Long,
    val time: Long,
    val title: String,
    val detail: String,
    val isError: Boolean,
)

data class UserNotice(
    val title: String,
    val detail: String,
    val isError: Boolean,
)

data class HandoffProgress(
    val kind: String,
    val title: String,
    val detail: String,
    val step: Int,
    val active: Boolean,
)

fun normalizeAddress(value: String?): String? {
    if (value.isNullOrBlank()) return null
    val compact = value.trim().replace(":", "").replace("-", "")
    if (compact.length != 12 || compact.any { it !in '0'..'9' && it !in 'a'..'f' && it !in 'A'..'F' }) return null
    return compact.uppercase(Locale.US).chunked(2).joinToString(":")
}

fun sameAddress(a: String?, b: String?): Boolean =
    normalizeAddress(a) != null && normalizeAddress(a) == normalizeAddress(b)

fun noiseLabel(value: String?): String = when (value) {
    "normal" -> "关闭"
    "reduction" -> "降噪"
    "ambient" -> "通透"
    else -> value ?: "未知"
}

fun effectLabel(value: String?): String = when (value) {
    "normal" -> "标准"
    "pop" -> "流行"
    "classical" -> "古典"
    "rock" -> "摇滚"
    else -> value ?: "未知"
}

fun ldacLabel(value: String?): String = when (value) {
    "off" -> "关闭"
    "rate48k" -> "44.1k / 48k"
    "rate96k" -> "96k"
    else -> value ?: "未知"
}

File: apps/windows-winui/Native/SessionModels.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace EdifierCtrl.Native;

[Microsoft.UI.Xaml.Data.Bindable]
public sealed record HeadphoneProfile
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = "";
    [JsonPropertyName("display_name")]
    public string DisplayName { get; set; } = "";
    [JsonPropertyName("features")]
    public IReadOnlyList<string> Capabilities { get; set; } = [];
    [JsonPropertyName("max_name_len")]
    public int MaxNameLen { get; set; }
    [JsonPropertyName("unique_service_uuid")]
    public string? ServiceUuid { get; set; }

    public bool Supports(string feature) => Capabilities.Contains(feature, StringComparer.Ordinal);
}

public sealed record HeadphoneDevice
{
    [JsonPropertyName("address")]
    public string Address { get; init; } = "";
    [JsonPropertyName("name")]
    public string Name { get; init; } = "";
    [JsonPropertyName("kind")]
    public string Kind { get; init; } = "rfcomm";
    [JsonPropertyName("service_uuid")]
    public string? ServiceUuid { get; init; }
    [JsonIgnore]
    public string? PeerId { get; init; }
    [JsonIgnore]
    public string DisplayName => string.IsNullOrWhiteSpace(Name) ? "未命名耳机" : Name;
    [JsonIgnore]
    public string ActionText { get; init; } = "连接控制";
}

public sealed record GroupPeer
{
    [JsonPropertyName("id")]
    public string Id { get; init; } = "";
    [JsonPropertyName("hostname")]
    public string Host { get; init; } = "";
    [JsonPropertyName("holding")]
    public string? Holding { get; init; }
    [JsonPropertyName("os")]
    public string Platform { get; init; } = "";
    [JsonPropertyName("app_version")]
    public string AppVersion { get; init; } = "";
    [JsonPropertyName("can_audio")]
    public bool CanAudio { get; init; }
    [JsonIgnore]
    public string DisplayName => string.IsNullOrWhiteSpace(Host) ? Id : Host;
}

public sealed record ActivityEntry(DateTimeOffset Time, string Title, string Detail, bool IsError);

internal sealed record NativeSnapshot(IReadOnlyList<JsonElement> Events, IReadOnlyList<GroupPeer>? Peers, string? Holding);

internal static class NativeJson
{
    internal static T Decode<T>(string json) => JsonSerializer.Deserialize<T>(json)
        ?? throw new JsonException("耳机服务返回了空数据.");

    internal static string? Text(JsonElement value, string key) =>
        value.TryGetProperty(key, out var item) && item.ValueKind == JsonValueKind.String ? item.GetString() : null;

    internal static bool? Boolean(JsonElement value, string key) =>
        value.TryGetProperty(key, out var item) && item.ValueKind is JsonValueKind.True or JsonValueKind.False ? item.GetBoolean() : null;

    internal static int? Integer(JsonElement value, string key) =>
        value.TryGetProperty(key, out var item) && item.ValueKind == JsonValueKind.Number && item.TryGetInt32(out var number) ? number : null;
}
